cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+   * over the `AWS_ENDPOINT_URL` environment variable.
+   */
+  S3Context(const std::string& endpoint_override = "") : _shutdown_s3_api{true}
+  {
+    // NB: `Aws::InitAPI` has to be called before everything in the SDK beside `Aws::SDKOptions`,
+    // even before config structs like `Aws::Client::ClientConfiguration`.
+    // Notice, we may call `Aws::InitAPI` multiple times, which is allowed see:
+    // <https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/Aws.cpp#L32>
+    Aws::SDKOptions options;
+    Aws::InitAPI(options);
+
+    // Create a client config where `endpoint_override` takes precedences over `AWS_ENDPOINT_URL`
+    Aws::Client::ClientConfiguration config;
+    const char* ep = std::getenv("AWS_ENDPOINT_URL");
+    if (!endpoint_override.empty()) {
+      config.endpointOverride = endpoint_override;
+    } else if (ep != nullptr && !std::string(ep).empty()) {
+      config.endpointOverride = ep;
+    }
+
+    // We check authentication here to trigger an early exception.
+    Aws::Auth::DefaultAWSCredentialsProviderChain provider;
+    if (provider.GetAWSCredentials().IsEmpty()) {
+      throw std::runtime_error("failed authentication to S3 server");
+    }
+    _client = std::make_shared<Aws::S3::S3Client>(config);
+  }
+
+  ~S3Context() noexcept
+  {
+    if (_shutdown_s3_api) {
+      try {
+        Aws::SDKOptions options;
+        Aws::ShutdownAPI(options);
+      } catch (const std::exception& e) {
+        std::cerr << "~S3Context(): " << e.what() << std::endl;
+      }
+    }
+  }
+
+  /**
+   * @brief Get a reference to the S3 client.
+   *
+   * @return S3 client.
+   */
+  Aws::S3::S3Client& client() { return *_client; }
+
+  // No copy and move semantic
+  S3Context(S3Context const&)       = delete;
+  void operator=(S3Context const&)  = delete;
+  S3Context(S3Context const&&)      = delete;
+  void operator=(S3Context const&&) = delete;
+
+  /**
+   * @brief Get the size of a S3 file
+   *
+   * @param bucket_name The bucket name.
+   * @param object_name The object name.
+   * @return Size of the file in bytes.
+   */
+  std::size_t get_file_size(const std::string& bucket_name, const std::string& object_name)
+  {
+    KVIKIO_NVTX_FUNC_RANGE();
+    Aws::S3::Model::HeadObjectRequest req;
+    req.SetBucket(bucket_name.c_str());
+    req.SetKey(object_name.c_str());
+    Aws::S3::Model::HeadObjectOutcome outcome = client().HeadObject(req);
+    if (!outcome.IsSuccess()) {
+      const Aws::S3::S3Error& err = outcome.GetError();
+      throw std::invalid_argument("get_file_size(): " + err.GetExceptionName() + ": " +
+                                  err.GetMessage());
+    }
+    return outcome.GetResult().GetContentLength();
+  }
+};
+
+/**
+ * @brief Handle of remote file (currently, only AWS S3 is supported).
+ */
+class RemoteHandle {
+ private:
+  std::string _bucket_name{};
+  std::string _object_name{};
+  std::size_t _nbytes{};
+  std::shared_ptr<S3Context> _context;
```

Comment:
Can this be a non-owning `S3Context const & _context;` or does that cause too many problems?

Reply:
I think we can, it looks like all the relevant methods of `Aws::S3::S3Client ` is `const`.
But I think it is a bit too dangerous, particularly, for the Python crowd.